content/posts/2024-08-29-gsoc-three-way-comparison.md

```diff
@@ -0,0 +1,90 @@
+
+# Background
+
+Three-way comparison is an operation present in many high-level languages, such as C++ and its starhip operator or Rust and the `Ord` trait. It operates on two values for which there is a defined comparison operation and returns `-1` if the first operand is less than the second, `0` if they are equal, and `1` otherwise. At the moment, compilers that use LLVM express this operation using different sequences of instructions which are optimized and lowered individually rather than as a single operation. Adding an intrinsic for this operation would therefore help us generate better machine code on some targets, as well as potentially optimize patterns in the middle-end that we didn't optimize before.
+
+# What was done
+
+Over the course of the project I have added two new intrinsics to the LLVM IR: `llvm.ucmp` for an unsigned 3-way comparison and `llvm.scmp` for a signed comparison. They both take two arguments that must be integers or vectors of integers and return an integer or a vector of integers with the same number of elements. The  arguments and the result do not need to have the same type.
+
+In the middle-end the following passes received some support for these intrinsics:
+
+  * InstSimplify ([#1](https://github.com/llvm/llvm-project/pull/93730), [#2](https://github.com/llvm/llvm-project/pull/95601))
+  * InstCombine ([#1](https://github.com/llvm/llvm-project/pull/96118), [#2](https://github.com/llvm/llvm-project/pull/98360), [#3](https://github.com/llvm/llvm-project/pull/101049), [#4](https://github.com/llvm/llvm-project/pull/105272), [#5](https://github.com/llvm/llvm-project/pull/105583))
+  * [CorrelatedValuePropagation](https://github.com/llvm/llvm-project/pull/97235)
+  * [ConstraintElimination](https://github.com/llvm/llvm-project/pull/97974)
+
+I have also added folds of idiomatic ways that a 3-way comparison can be expressed to a call to the corresponding intrinsic.
+
+In the backend there are two different ways of expanding the intrinsics: [as a nested select](https://github.com/llvm/llvm-project/pull/91871) (i.e. `(x < y) ? -1 : (x > y ? 1 : 0)`) or [as a subtraction of zero-extended comparisons](https://github.com/llvm/llvm-project/pull/98774) (`zext(x > y) - zext(x < y)`). The first option is the default one, but targets can choose to use the second one through a TLI hook.
+
+# Results
+
+I think that overall the project was succesful and brought a small positive change to LLVM. To demonstrate its impact in a small test case, the following function in C was compiled twice, first with Clang 18.1 and then with Clang built from the main branch of LLVM repository:
+
+```C
+unsigned char ucmp_32_8(unsigned int a, unsigned int b) {
+    return (a < b ? -1 : a > b);
+}
```

Comment:
I was going to say that we would get an even more compelling example by using the spaceship operator, because it generates pretty terrible asm on 18.1 (8 instructions instead of 3): https://c.godbolt.org/z/WGP7vba14

But we need https://github.com/llvm/llvm-project/pull/107314 to land before we actually get the good new codegen for this pattern :)

Reply:
The patch has landed, so now this example goes from
```
spaceship(unsigned int, unsigned int):
        xor     ecx, ecx
        cmp     edi, esi
        mov     eax, 0
        sbb     eax, eax
        or      al, 1
        cmp     edi, esi
        movzx   eax, al
        cmove   eax, ecx
        ret
```
to
```
spaceship(unsigned int, unsigned int):
        cmp     edi, esi
        seta    al
        sbb     al, 0
        ret
```
which is quite nice!